src/Cardano/Wallet/AddressDerivation.hs

```diff
@@ -0,0 +1,304 @@
+    , XPrv
+    , XPub
+    , deriveXPrv
+    , deriveXPub
+    , generateNew
+    , toXPub
+    )
+import Cardano.Wallet.Binary
+    ( encodeAddress )
+import Cardano.Wallet.Primitive
+    ( Address (..) )
+import Control.DeepSeq
+    ( NFData )
+import Data.ByteArray
+    ( ScrubbedBytes )
+import Data.ByteString
+    ( ByteString )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Word
+    ( Word32 )
+import GHC.Generics
+    ( Generic )
+import GHC.TypeLits
+    ( Symbol )
+
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Write as CBOR
+
+
+{-------------------------------------------------------------------------------
+                        Polymorphic / General Purpose Types
+-------------------------------------------------------------------------------}
+
+-- | A cryptographic key, with phantom-types to disambiguate key types.
+--
+-- @
+-- let rootPrivateKey = Key 'RootK XPrv
+-- let accountPubKey = Key 'AccountK XPub
+-- let addressPubKey = Key 'AddressK XPub
+-- @
+newtype Key (level :: Depth) key = Key key
+    deriving stock (Generic, Show, Eq)
+
+instance (NFData key) => NFData (Key level key)
+
+
+-- | Key Depth in the derivation path, according to BIP-0039 / BIP-0044
+--
+-- root' / purpose' / cointype' / account' / change / address
+--
+-- We do not manipulate purpose, cointype and change paths directly, so they are
+-- left out of the sum type.
+data Depth = RootK | AccountK | AddressK
+
+-- | A derivation index, with phantom-types to disambiguate derivation type.
+--
+-- @
+-- let accountIx = Index 'Hardened 'AccountK
+-- let addressIx = Index 'Soft 'AddressK
+-- @
+newtype Index (derivationType :: DerivationType) (level :: Depth) = Index
+    { getIndex :: Word32 }
+    deriving stock (Generic, Show, Eq, Ord)
+
+instance NFData (Index derivationType level)
+
+instance Bounded (Index 'Hardened level) where
+    minBound = Index 0x80000000
+    maxBound = Index maxBound
+
+instance Bounded (Index 'Soft level) where
+    minBound = Index minBound
+    maxBound = let (Index ix) = minBound @(Index 'Hardened _) in Index (ix - 1)
+
+instance Enum (Index 'Hardened level) where
+    fromEnum (Index ix) = fromIntegral ix
+    toEnum ix
+        | Index (fromIntegral ix) < minBound @(Index 'Hardened _) =
+            error "Index@Hardened.toEnum: bad argument"
+        | otherwise =
+            Index (fromIntegral ix)
+
+instance Enum (Index 'Soft level) where
+    fromEnum (Index ix) = fromIntegral ix
+    toEnum ix
+        | Index (fromIntegral ix) > maxBound @(Index 'Soft _) =
+            error "Index@Soft.toEnum: bad argument"
+        | otherwise =
+            Index (fromIntegral ix)
+
+
+-- | Type of derivation that should be used with the given indexes.
+data DerivationType = Hardened | Soft
+
+-- | An encapsulated passphrase. The inner format is free, but the wrapper helps
+-- readability in function signatures.
+newtype Passphrase (goal :: Symbol) = Passphrase ScrubbedBytes
```

Comment:
just wonder why for `Index` and `Key` there is `Depth` and `DerivationType` phantom-type used, respectively. And here, you are relying on Symbol (in a sense making phantom-type like specialization open) rather than on **data**.  Using : 
```
data PassphrasePurpose = Encryption | Generation
```
would have some drawbacks? Maybe we want to extend it somewhere outside ant making not visible in this module?

Reply:
Exactly for the reason you mention, using a symbol leaves the type quite open which makes sense for types that are rather general like "Passphrase". On the contrary, the Index and Key have a very precise semantic for which  we already know the spectrum, and therefore, we can be explicit and have a closed typed.

So, it's just slightly more convenient to have symbols than sum types and we get almost the same guarantee. Typos are still possible but in such case, GHC will probably yell at you that types don't match.

Reply:
:+1: